### ISSUE TITLE
podman save: cannot handle colons in output filename

### ISSUE THREAD

**Issue by edsantiago** [projectatomic/libpod#489]
If the `--output` path includes a colon, `podman save` fails with an unhelpful diagnostic:

    # podman save --output /tmp/a:b docker.io/fedora:latest
    ERRO[0000] error deleting "/tmp/a:b": error getting destination imageReference for "docker-archive:/tmp/a:b:docker.io/fedora:latest": invalid reference format 
    unable to save "docker.io/fedora:latest": error getting destination imageReference for "docker-archive:/tmp/a:b:docker.io/fedora:latest": invalid reference format

Possible hint: if the left-hand side of the colon includes any capital letter, the diagnostic is:

    # podman save --output /tmp/A:b docker.io/fedora:latest
    ERRO[0000] error deleting "/tmp/A:b": error getting destination imageReference for "docker-archive:/tmp/A:b:docker.io/fedora:latest": invalid reference format: repository name must be lowercase 
    unable to save "docker.io/fedora:latest": error getting destination imageReference for "docker-archive:/tmp/A:b:docker.io/fedora:latest": invalid reference format: repository name must be lowercase

Colons are vaild path components; a real-world example in which one might want colons is using the output of `podman images` (sha256:xxxxx) to generate a filename.

**Comment by umohnani8** [comment]
@edsantiago it is working for me.
```
➜  libpod git:(master) ✗ sudo podman save -o blah:bre alpine
Getting image source signatures
Copying blob sha256:cd7100a72410606589a54b932cabd804a17f9ae5b42a1882bd56d263e02b6215
 4.20 MB / 4.20 MB [========================================================] 0s
Copying config sha256:3fd9065eaf02feaf94d68376da52541925650b81698c53c6824d92ff63f98353
 1.48 KB / 1.48 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
➜  libpod git:(master) ✗ sudo podman save -o /tmp/blah:bre alpine
Getting image source signatures
Copying blob sha256:cd7100a72410606589a54b932cabd804a17f9ae5b42a1882bd56d263e02b6215
 4.20 MB / 4.20 MB [========================================================] 0s
Copying config sha256:3fd9065eaf02feaf94d68376da52541925650b81698c53c6824d92ff63f98353
 1.48 KB / 1.48 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
➜  libpod git:(master) ✗ sudo podman save -o /tmp/Blah:bre alpine
Getting image source signatures
Copying blob sha256:cd7100a72410606589a54b932cabd804a17f9ae5b42a1882bd56d263e02b6215
 4.20 MB / 4.20 MB [========================================================] 0s
Copying config sha256:3fd9065eaf02feaf94d68376da52541925650b81698c53c6824d92ff63f98353
 1.48 KB / 1.48 KB [========================================================] 0s
Writing manifest to image destination
Storing signatures
```
Looks like the issue is with the image name. It should be `docker.io/library/fedora` and not `docker.io/fedora`.
The error is definitely misleading.

**Comment by umohnani8** [comment]
Just found another issue, if the file is `blah:bloo.tar`, it saves it as just `blah`. Not sure why that is happening...

**Comment by edsantiago** [comment]
Hm, no, something changed in podman-0.3.3-2.gitb47fc0d.fc27. I can no longer reproduce the original error. I do see the new symptom you report (stripping the colon and everything after).

**Comment by umohnani8** [comment]
@baude you reworte the way podman handles images yes? Does something come to mind that could be causing this?

**Comment by mheon** [comment]
Probably c/image changes, we updated our vendored version for 0.3.3

**Comment by baude** [comment]
No its more likely me

**Comment by baude** [comment]
@edsantiago Ill follow up with you

**Comment by baude** [comment]
The culprit appears to be https://github.com/projectatomic/libpod/blob/master/libpod/runtime_img.go#L740 where the user's input is used to create a destination for copy.  I'll have to check with mitr if he thinks that transport should handle it.

**Comment by rhatdan** [comment]
@mitrmac ^^

**Comment by umohnani8** [comment]
@mtrmac ^^

**Comment by mtrmac** [comment]
> Just found another issue, if the file is `blah:bloo.tar`, it saves it as just blah. Not sure why that is happening...

Because https://github.com/projectatomic/libpod/blob/master/cmd/podman/save.go#L120 constructs a `docker-archive:blah:bloo.tar:`_image_, where _image_ is an unmodified string from the CLI, and as described above, this is parsed as
- transport `docker-archive`
- path `blah`
- reference
  - name `bloo.tar`
  -  tag _image_

as long as _image_ is a valid Docker-like tag.

It would be nice if `podman` instituted a rule against ever using string concatenation for reference names in the future.

(BTW https://github.com/projectatomic/libpod/blob/master/cmd/podman/save.go#L119 _really_ should be a `HasPrefix`, not `Contains`! Strings, again.)

**Comment by mtrmac** [comment]
```console
# podman save --output /tmp/a:b docker.io/fedora:latest
ERRO[0000] error deleting "/tmp/a:b": error getting destination imageReference for "docker-archive:/tmp/a:b:docker.io/fedora:latest": invalid reference format 
unable to save "docker.io/fedora:latest": error getting destination imageReference for "docker-archive:/tmp/a:b:docker.io/fedora:latest": invalid reference format
```

The duplicate “error deleting” is https://github.com/projectatomic/libpod/blob/master/cmd/podman/save.go#L124 using `err` instead of `err2`.  Maybe `err2` should be silently ignored `if os.IsNotExist(err2)`.

The `invalid reference format` is fundamentally caused by parsing the path as `/tmp/a`, as described above. The exact flow seems to be (reading, not testing the code):

- https://github.com/projectatomic/libpod/blob/master/libpod/runtime_img.go#L740 fails with `docker-archive parsing reference: invalid reference format`, which is discarded
- `hasTransport` returns `false` (no `://`)
- https://github.com/projectatomic/libpod/blob/master/libpod/runtime_img.go#L747 and https://github.com/projectatomic/libpod/blob/master/libpod/runtime_img.go#L751 both try to parse `docker://docker-archive:/tmp/a:b:docker.io/fedora:latest`, which fails with a context-less `invalid reference format`.

---

Overall, it would be helpful if all parsing errors in c/image made sure to quote their exact input, to make tracing the strings easier.

As for the destination parsing heuristics in `Runtime.PushImage`, pedantically I’d prefer for it not to exist at all; but given that it does, it’s not obvious which of the three different error messages should be returned. In this case it would be much more helpful to return the error from the first `ParseImageName` (which contains an almost-correctly-formed `docker-archive:` reference), instead of the error from the `docker://docker-archive:…` input, but if the user does not specify a reference and expects Docker semantics, the last error is more useful.

**Comment by mtrmac** [comment]
```console
# podman save --output /tmp/A:b docker.io/fedora:latest
ERRO[0000] error deleting "/tmp/A:b": error getting destination imageReference for "docker-archive:/tmp/A:b:docker.io/fedora:latest": invalid reference format: repository name must be lowercase 
unable to save "docker.io/fedora:latest": error getting destination imageReference for "docker-archive:/tmp/A:b:docker.io/fedora:latest": invalid reference format: repository name must be lowercase
```
Same flow as above, the error is from parsing `docker://docker-archive:/tmp/A:b:docker.io/fedora:latest` (where the `c/i/docker/reference` parser heuristically guesses that upper-case is the most likely / most important aspect of an invalid format).

**Comment by mtrmac** [comment]
> Hm, no, something changed in podman-0.3.3-2.gitb47fc0d.fc27. I can no longer reproduce the original error. I do see the new symptom you report (stripping the colon and everything after).

Is that really the case? I can’t quickly see a notable change in implementation between the 0.3.3 and 0.3.2 releases. Am I missing something, or was the behavior change really driven by a change in the test? (Note that the “source” changes the `docker-archive:…` string created for the destination, i.e. also the behavior in some cases.)

**Comment by edsantiago** [comment]
>>Hm, no, something changed in podman-0.3.3-2.gitb47fc0d.fc27. I can no longer reproduce the original error. I do see the new symptom you report (stripping the colon and everything after).
>
>Is that really the case?

No. My bad. My "something changed" was that my host had rebooted, and I had forgotten to unset XDG_WHATEVER. Here is a consistent reproducer that also addresses the discrepancy @umohnani8 notes between `fedora` and `library/fedora`:
```
# unset XDG_RUNTIME_DIR                    ! causes weird failures if set
# podman pull docker.io/fedora
# podman save --output /tmp/a:b docker.io/fedora:latest
[output as described in initial post]
```
I will add that (1) docker handles this just fine, and (2) it doesn't seem to make much sense (to me) to treat the output file of `--save` as a possible image. Even the man page explicitly states that it is a file.

I am sorry for the confusion. I hope this reproducer is more, um, reproducible.

**Comment by mtrmac** [comment]
> (2) it doesn't seem to make much sense (to me) to treat the output file of `--save` as a possible image.

Sure; this is a bug caused by combining various components into a single string and then extracting the components again. We should definitely, at least, refuse such input if we can’t handle it.

I think it would be cleanest to replace https://github.com/projectatomic/libpod/blob/master/cmd/podman/save.go#L102 + https://github.com/projectatomic/libpod/blob/master/cmd/podman/save.go#L120 with a call to a new `c/i/docker/archive.NewReference(path, parsedReference)`, where `parsedReference` is a typed `reference.NamedTagged`, either by parsing `image` separately, or maybe preferably by https://github.com/projectatomic/libpod/blob/master/libpod/runtime_img.go#L772 + `DockerReference`; and then pass the resulting `types.ImageReference` through to `PushImage` without calling `ParseImageName` on it at all.  Admittedly a rather extensive restructuring.

As for actually handling it, it’s possible but awkward — see the earlier discussion about escaping.  (One nice thing about the `NewReference` approach above is that if `docker-archive:` did change the separator or introduced escaping, the `NewReference` caller would not need to change at all—only more `path` values would be accepted.)

(There are even more gross solutions, e.g. create a `/tmp/output`_$random_ symlink which contains the original path; then the `docker-archive:`_path_ input will not need to include the metacharacters.)

**Comment by mheon** [comment]
It sounds like the easiest solution here would be to declare `:` a reserved character in filenames created by `podman save`, to prevent ambiguities in parsing generated references. I have no issues with this approach, as I doubt many people include `:` in filenames anyways, and files can be trivially renamed if a `:` is truly necessary.

Does this also break `podman import`?

**Comment by mtrmac** [comment]
> Does this also break `podman import`?

AFAICT it does (`tarball:` accepts a colon-separated list of tarballs).  `podman load` as well (via the same `docker-archive`/`oci-archive` transports as `save`).

**Comment by mheon** [comment]
Alright. It seems like it would make sense to have a uniform restriction that filenames handled by `import`, `load`, and `save` cannot contain `:` characters.

**Comment by edsantiago** [comment]
>I doubt many people include : in filenames 

More than you might think. `date --iso-8601=seconds` is a pretty common convention for generating filenames, and it emits colons in the HH:MM:SS part. That's not to say you can't block colons, just that it will irk some people.

**Comment by rhatdan** [comment]
Lets do it for now and wait for some one to actually complain on this.

**Comment by umohnani8** [comment]
Has this been fixed, or do we still need to document that `:` is restricted character for file names?

**Comment by umohnani8** [comment]
@mheon thanks for the info! Working on it.

**Comment by umohnani8** [comment]
Fixed in #694
